### PR TITLE
[JIT] Enable freezing for sparse COO tensors

### LINKED ISSUE #63987: [JIT] Serialization of Sparse Tensors doesnt work
## 🐛 Bug
```
import torch
import io

class SparseTensorModule(torch.nn.Module):
	def __init__(self):
		super().__init__()
		self.a = torch.sparse.FloatTensor()

	def forward(self):
		pass

torch.jit.save(torch.jit.script(SparseTensorModule()), io.BytesIO())
```

> NotImplementedError: Cannot access storage of SparseTensorImpl

## Additional context

Ran into this when fixing a freezing / sparsity bug which was requested internally.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69614

Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

File: aten/src/ATen/core/ivalue.h [hunk 1]
```diff
@@ -291,16 +309,7 @@ struct TORCH_API IValue final {
...
-      const auto& rhsTensor = rhs.toTensor();
-      // mkldnn tensors dont have views or storage, so we compare
-      // based on tensor impl. //TODO: find a way to use mkldnn storage
-      if (thisTensor.is_mkldnn() || rhsTensor.is_mkldnn()) {
-        return thisTensor.unsafeGetTensorImpl() ==
-            rhsTensor.unsafeGetTensorImpl();
-      }
-
-      return thisTensor.is_alias_of(rhsTensor);
+      return isAliasOf(this->toTensor(), rhs.toTensor());
```
Comment (eellison) [inline aten/src/ATen/core/ivalue.h:312]: nice refactoring!

File: aten/src/ATen/core/ivalue.h [hunk 2]
```diff
@@ -872,17 +881,22 @@ struct TORCH_API IValue final {
...
   // Detect aliased tensors.
   struct HashAliasedIValue {
+    size_t hashTensor(const at::Tensor& ten) const {
+      if (ten.is_mkldnn()) {
+        // MKLDNN tensors dont have storage and dont create views
+        // or aliasing so we can just use Tensor pointer, TODO: find way
+        // to use mkldnn storage
+        return reinterpret_cast<size_t>(ten.unsafeGetTensorImpl());
+      } else if (ten.is_sparse()) {
+        return hashTensor(ten._values());
```
Comment (eellison) [inline aten/src/ATen/core/ivalue.h:891]: so, our options are:
- hash on just index/values
this makes it so we will detect aliasing overlap between sparse tensors and non-sparse tensors only of one of the two.  
- maintain separate hashes for index and value... i think this is probably not worth it for what is the extremely rare (non-existant ?) case of mutating after freezing sparse tensors indices that also happens to be a tensor that is used elsewhere
- 
i think current approach is good, maybe we can  add comment re-ignoring indices

File: test/jit/test_sparse.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    def test_serialize_sparse_coo(self):
+        class SparseTensorModule(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.a = torch.rand(3, 4).to_sparse()
+
+            def forward(self, x):
+                return x + self.a
+
+        torch.jit.save(torch.jit.script(SparseTensorModule()), io.BytesIO())
```
Comment (eellison) [inline test/jit/test_sparse.py:39]: can we test load as well? there are examples in test_jit of how to do that

File: test/jit/test_sparse.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+        x = torch.rand(3, 4).to_sparse()
+
+        m = SparseTensorModule()
+        unfrozen_result = m.forward(x)
+
+        m.eval()
+        frozen = torch.jit.freeze(torch.jit.script(m))
+
+        frozen_result = frozen.forward(x)
```
Comment (eellison) [inline test/jit/test_sparse.py:26]: can we test save-load here just to be safe ?

**Review by eellison — APPROVED** [review]
Nice!! couple of small comments

**Comment by davidberard98** [comment]
Turns out there was still an issue with freezing/serializing modules that had more than 1 sparse tensor, updated to fix that

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

**Pull request closed without merging**